src/components/position_calc/position_calc_utils.py

```diff
@@ -0,0 +1,82 @@
+import numpy as np
+from scipy import optimize
+from simulator_main import sim_config as cfg
+from sim_utils.common_types import *
+
+# TODO: documentation (RIP)
+def gradient_descent(func, starting_params, args=(), step_size=0.5, termination_ROC=1e-5, 
+                        max_iter=1e5, is_grad=False, delta_x=0.01):
+    params = starting_params
+    gradient = get_grad(func, params, args, is_grad, delta_x)
+    grad_mag = np.linalg.norm(gradient)
+    num_iter = 0
+    print(gradient)
+
+    while (grad_mag > termination_ROC):
+        params = params - step_size*gradient
+        gradient = get_grad(func, params, args, is_grad, delta_x)
+        grad_mag = np.linalg.norm(gradient)
+        num_iter += 1
+
+        if (grad_mag == np.inf):
+            print("OverflowError. Step size might be too big. Optimization diverges")
+            print("Diverging Parameters: ", params)
+            return params
+
+        if (num_iter == max_iter):
+            print("WARNING! maximum iteration number reached")
+            break
+
+    print("Gradient Descent with step-size %f finished after %0d iterations" %(step_size, num_iter))
+    return params
+
+def nelder_mead(func, starting_params, args=()):
+    results = optimize.minimize(func, starting_params, args=args, method='Nelder-Mead')
+
+    if (not results.success):
+        print(results.message)
+
+    # print("Converged with %d iterations" %results.nit)
+    # print(func(results.x, *args))
+
+    return results.x
+
+def newton_gauss(func, starting_params, args=()):
+    pass
+
+def get_grad(func, params, args, is_grad, delta_x):
+    if (is_grad):
+        return func(params, *args)
+    else:
+        return np.array([get_derivative(func, params, args, i, delta_x)
+                        for i in range(len(params))])
+
+def get_derivative(func, params, args, var_index, delta_x):
+    delta_params = np.zeros(params.shape)
+    delta_params[var_index] = delta_x
+
+    return (func(params + delta_params, *args) - func(params, *args))/delta_x
+
+def tdoa_function_3D(pinger_pos, hydrophone_pos, is_polar):
```

Comment:
This probably should have a different name since it doesn't actually calculate tdoa right? It calculates time of arrival for a given hydrophone? Maybe time_of_arrival_for_hydrophone() or something?

Reply:
TDOA = time difference of arrival I was just referring to the wrong thing as tdoa and decided to fix it haha

Reply:
I mean I thought this doesn't calculate time difference
Just time right?
Wait does this mean time difference from pinger to hydrophone?
I thought time difference was for hydrophone to other hydrophone